Kotlin immutable collections benchmarks

qurbonzoda opened the pull request.

ilya-g commented on benchmarks-mpp/gradle/wrapper/gradle-wrapper.properties:
```diff
@@ -0,0 +1,5 @@
+distributionBase=GRADLE_USER_HOME
```
This file is unneeded because there's no standalone wrapper in this directory.

qurbonzoda replied:
deleted

ilya-g commented on benchmarks-mpp/build.gradle:
```diff
@@ -0,0 +1,39 @@
+plugins {
+    id 'org.jetbrains.kotlin.multiplatform' version '1.3.31'
+    id 'org.jetbrains.gradle.benchmarks.plugin' version '0.1.7-dev-21'
```
Try using 0.2.0-dev-2 from kotlinx repository: https://bintray.com/kotlin/kotlinx/kotlinx.benchmark

qurbonzoda replied:
+1

ilya-g commented on benchmarks-runner/src/main/kotlin/runners/hashMapRunner.kt:
```diff
@@ -0,0 +1,30 @@
+
+fun main(args: Array<String>) {
+    runBenchmarks(hashMapOutputFileName) { this
+            .include("immutableMap.Put.put$")
```
Is it left here accidentally?

qurbonzoda replied:
yes, it is

ilya-g commented on benchmarks-runner/src/main/kotlin/runners/hashMapBuilderRunner.kt:
```diff
@@ -0,0 +1,29 @@
+import runBenchmarks
+
+
+fun main(args: Array<String>) {
```
Minor: args are not used in runners so they can be omitted.

qurbonzoda replied:
done

ilya-g commented on benchmarks-mpp/src/jvmMain/kotlin/benchmarks/immutableList/Add.kt:
```diff
@@ -0,0 +1,55 @@
+@Warmup(iterations = 5)
+@Measurement(iterations = 5)
+@BenchmarkMode(Mode.AverageTime)
+@OutputTimeUnit(TimeUnit.MICROSECONDS)
```
Are all these annotations needed, if we override them from both `benchmarks-mpp` and `benchmarks-runner`?

qurbonzoda replied:
`benchmarks-runner` overrode only warmup and measurement iterations. Now it override benchmark mode and output time unit as well.
So, these annotations are removed now.

ilya-g commented on benchmarks-runner/src/main/kotlin/regressionCalculator.kt:
```diff
@@ -0,0 +1,76 @@
+        if (reference != null) {
+            regressScore = target.score - reference.score
+            regressScorePercent = 100 * regressScore / reference.score
+            regressScoreError = target.scoreError + reference.scoreError
```
I believe it would be more correct to calculate the error of difference as `sqrt(err1^2 + err2^2)`

qurbonzoda replied:
fixed

ilya-g commented:
Let's squash some commits and rebase on master (base branch of this PR can be changed as well) to continue the review.

qurbonzoda commented:
Rebased on master and squashed some commits, as you asked.
Commits are displayed on GitHub in a different order. It is because commit time has changed for some of them due to squash/interactive rebase.

ilya-g commented on benchmarks-runner/build.gradle:
```diff
@@ -0,0 +1,140 @@
+    runtimeOnly project(path: ':kotlinx-collections-immutable')
+}
+
+task copyBenchmarkList(type: Copy, dependsOn: ':benchmarks-mpp:jvmBenchmarkGenerate', group: "Benchmark") {
```
Not holding the merge, but we need to investigate whether can we avoid copying this list and instead add that resources dir to the classpath.

qurbonzoda replied:
Ilya Ryzhenkov just helped me with this issue

ilya-g commented on benchmarks-runner/build.gradle:
```diff
@@ -0,0 +1,131 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
```
We usually don't place copyrights in build scripts

qurbonzoda replied:
+1
